File: backend/src/groots/domain/model/track.py
from dataclasses import dataclass, field
from datetime import datetime

from groots.domain.model.base import new_id


@dataclass
class Track:
    cid: str  # IPFS Content Identifier
    title: str
    duration_seconds: int
    file_size_bytes: int
    album_id: str | None = None  # Reference to Album entity (required for all tracks)
    id: str = field(default_factory=new_id)
    track_number: int | None = None
    mime_type: str = "audio/mpeg"
    pinned: bool = False  # True when central node has pinned the CID
    fingerprint_id: str | None = None  # ID of the TrackFingerprint record for this track
    matched_central_id: str | None = None  # fingerprint ID of central match, if any
    source: str | None = None  # e.g. "cd", "vinyl", "digital_download", "streaming_purchase"
    disc_number: int | None = None  # which disc/vinyl (1, 2…); null means single-disc album
    side: str | None = None  # vinyl side ("A", "B", "C", "D"); null for CDs
    created_at: datetime = field(default_factory=datetime.utcnow)

File: groots_app/lib/src/domain/models/track.dart
class Track {
  final String id;
  final String cid;
  final String title;
  final String? albumId;
  final int? trackNumber;
  final int durationSeconds;
  final int fileSizeBytes;
  final String mimeType;
  final bool pinned;
  final String? matchedCentralId;
  final String? source;
  final int? discNumber;
  final String? side;

  const Track({
    required this.id,
    required this.cid,
    required this.title,
    this.albumId,
    this.trackNumber,
    required this.durationSeconds,
    required this.fileSizeBytes,
    required this.mimeType,
    required this.pinned,
    this.matchedCentralId,
    this.source,
    this.discNumber,
    this.side,
  });

  factory Track.fromJson(Map<String, dynamic> json) => Track(
        id: json['id'] as String,
        cid: json['cid'] as String,
        title: json['title'] as String,
        albumId: json['album_id'] as String?,
        trackNumber: json['track_number'] as int?,
        durationSeconds: json['duration_seconds'] as int,
        fileSizeBytes: json['file_size_bytes'] as int,
        mimeType: json['mime_type'] as String,
        pinned: json['pinned'] as bool,
        matchedCentralId: json['matched_central_id'] as String?,
        source: json['source'] as String?,
        discNumber: json['disc_number'] as int?,
        side: json['side'] as String?,
      );

  String get durationFormatted {
    final m = durationSeconds ~/ 60;
    final s = durationSeconds % 60;
    return '$m:${s.toString().padLeft(2, '0')}';
  }
}
